### PR TITLE
Add before stop hook

### PR DESCRIPTION
**Pull request by le-duane**
We'd like the ability to run scripts prior to shutting down our Postgres instances.

The two cases we have in mind are:
* In spite of following all best practices client-side, logical replication connections can sometimes hang the Postgres shutdown sequence. We'd like to sigterm any misbehaving logical replication connections which remain after x seconds. These will inevitably get killed anyway on master stop timeout.
* remove "role=master" label on current primary when not using k8s as DCS. Waiting until after Postgres fully stops can sometimes be too long for this.

#2596

### CODE REVIEW

**Comment by CyberDem0n** [comment]
Thanks for working on it @le-duane.

Here are my thoughts. Currently callbacks are supposed to be fired and forgot and they will be running in parallel with whatever Patroni does with Postgres.
In some cases they are indeed executed only after reaching certain stage:
- on_start is executed when Postgres starts accepting connections
- on_stop is executed when Postgres is already stopped

But, right after `on_stop` there could be `on_start` and therefore the newly executed callback usually cancels the one that already running (Patroni kills the process and its all children).

The `on_pre_stop` feels like a bit special. There is one very useful use-case when we would like to run it not as fire-and-forget, but actually wait for the result. More specifically I am talking about `pgbouncer PAUSE`.

Of course, after promoting the new node (or when Postgres is back running as a primary after a restart) we want to do `pgbouncer RESUME`, but this one we don't need to wait for.

Also, we already have `postgresql.pre_promote` hook, which exit code is actually being checked and `pg_ctl promote` isn't called if the exit code is different from 0.

Having said that I'm thinking that `on_pre_stop` should not be a callback, but something similar to `pre_promote`, because we may want to wait until it finished before executing `pg_ctl stop -mf`. And probably `before_stop` would be a bit better name.

**Comment by le-duane** [comment]
Thanks for the quick reply @CyberDem0n - I'll look into `pre_promote`

File: patroni/postgresql/__init__.py [hunk 1]
```diff
@@ -1024,6 +1027,20 @@ def _pre_promote(self):
...
+        """
+
+        cmd = self.config.get('before_stop')
+        if not cmd:
+            return True
+
+        ret = self.cancellable.call(shlex.split(cmd))
+        if ret is not None:
+            logger.info('before_stop script `%s` exited with %s', cmd, ret)
+
```
Comment (CyberDem0n) [inline patroni/postgresql/__init__.py:1042]:
```suggestion
        try:
            ret = self.cancellable.call(shlex.split(cmd))
            if ret is not None:
                logger.info('before_stop script `%s` exited with %s', cmd, ret)
        except Exception as e:
            logger.error('Exception when calling `%s`: %r', cmd, e)
```

**Comment by le-duane** [comment]
Thanks for the tips @CyberDem0n. I'm a lot happier that this is running synchronously now. 

I changed things a little so I could add a test.

File: patroni/postgresql/__init__.py [hunk 2]
```diff
@@ -1024,6 +1027,25 @@ def _pre_promote(self):
...
 
+    def _before_stop(self):
+        """
+        Synchronously run a script prior to stopping postgres
+        """
+
+        cmd = self.config.get('before_stop')
+        if not cmd:
+            return True
+        else:
```
Comment (CyberDem0n) [inline patroni/postgresql/__init__.py:1038]:
```suggestion
        if cmd:
```

File: tests/test_postgresql.py [hunk 3]
```diff
@@ -195,6 +196,17 @@ def test_stop(self, mock_is_running):
...
+        # Ensure before_stop script is called only when configured to
+        mock_postmaster.wait.side_effect = [psutil.TimeoutExpired(30), Mock()]
+        self.p.config._config['before_stop'] = ':'
+        self.p.stop(on_safepoint=mock_callback, stop_timeout=30)
+        self.assertTrue(self.p._do_before_stop.call_count == 1)
+        self.p._do_before_stop.reset_mock()
+        self.p.config._config.pop('before_stop')
+        mock_postmaster.wait.side_effect = [psutil.TimeoutExpired(30), Mock()]
+        self.p.stop(on_safepoint=mock_callback, stop_timeout=30)
+        self.assertTrue(self.p._do_before_stop.call_count == 0)
```
Comment (CyberDem0n) [inline tests/test_postgresql.py:208]:
```suggestion
        # Ensure before_stop script is called when configured to                                                                       
        self.p.config._config['before_stop'] = ':'                                                                                     
        mock_postmaster.wait.side_effect = [psutil.TimeoutExpired(30), Mock()]                                                         
        mock_cancellable_call.return_value = 0                                                                                         
        with patch('patroni.postgresql.logger.info') as mock_logger:                                                                   
            self.p.stop(on_safepoint=mock_callback, stop_timeout=30)                                                                   
            self.assertEqual(mock_logger.call_args[0], ('before_stop script `%s` exited with %s', ':', 0))                             
        mock_postmaster.wait.side_effect = [psutil.TimeoutExpired(30), Mock()]                                                         
        mock_cancellable_call.side_effect = Exception                                                                                  
        with patch('patroni.postgresql.logger.error') as mock_logger:                                                                  
            self.p.stop(on_safepoint=mock_callback, stop_timeout=30)                                                                   
            self.assertEqual(mock_logger.call_args_list[1][0][0], 'Exception when calling `%s`: %r')                                   
```

File: patroni/postgresql/__init__.py [hunk 4]
```diff
@@ -1024,6 +1027,25 @@ def _pre_promote(self):
...
+            self._do_before_stop(cmd)
+
+    def _do_before_stop(self, cmd):
```
Comment (CyberDem0n) [inline patroni/postgresql/__init__.py:1041]:
```suggestion
    def _do_before_stop(self, cmd: str) -> None:
```

**Review by CyberDem0n — COMMENTED** [review]
Please also update https://github.com/zalando/patroni/blob/ff6d728f07e9a378c595115dabe11b0ca8dba84b/patroni/validator.py#L757 as `Optional("before_strop"): str`

File: tests/test_postgresql.py [hunk 5]
```diff
@@ -195,6 +196,19 @@ def test_stop(self, mock_is_running):
...
+        with patch('patroni.postgresql.logger.info') as mock_logger:                                                                   
+            self.p.stop(on_safepoint=mock_callback, stop_timeout=30)                                                                   
+            self.assertEqual(mock_logger.call_args[0], ('before_stop script `%s` exited with %s', ':', 0))                             
+        mock_postmaster.wait.side_effect = [psutil.TimeoutExpired(30), Mock()]                                                         
+        mock_cancellable_call.side_effect = Exception                                                                                  
+        with patch('patroni.postgresql.logger.error') as mock_logger:                                                                  
+            self.p.stop(on_safepoint=mock_callback, stop_timeout=30)                                                                   
+            self.assertEqual(mock_logger.call_args_list[1][0][0], 'Exception when calling `%s`: %r')                                   
+
```
Comment (CyberDem0n) [inline tests/test_postgresql.py:211]:
```suggestion
        # Ensure before_stop script is called when configured to
        self.p.config._config['before_stop'] = ':'
        mock_postmaster.wait.side_effect = [psutil.TimeoutExpired(30), Mock()]
        mock_cancellable_call.return_value = 0
        with patch('patroni.postgresql.logger.info') as mock_logger:
            self.p.stop(on_safepoint=mock_callback, stop_timeout=30)
            self.assertEqual(mock_logger.call_args[0], ('before_stop script `%s` exited with %s', ':', 0))
        mock_postmaster.wait.side_effect = [psutil.TimeoutExpired(30), Mock()]
        mock_cancellable_call.side_effect = Exception
        with patch('patroni.postgresql.logger.error') as mock_logger:
            self.p.stop(on_safepoint=mock_callback, stop_timeout=30)
            self.assertEqual(mock_logger.call_args_list[1][0][0], 'Exception when calling `%s`: %r')

```

**Comment by hughcapet** [comment]
👍

**Comment by CyberDem0n** [comment]
:+1:

**Comment by CyberDem0n** [comment]
Merged, thank you @le-duane!

**Pull request merged**